sql/catalyst/src/main/java/org/apache/spark/sql/connector/CustomMetric.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.read.Scan;
+
+/**
+ * A custom metric. This is a logical representation of a metric reported by data sources during
+ * read path. Data sources can report supported metric list by {@link Scan} to Spark in query
+ * planning. During query execution, Spark will collect the metrics per partition by
+ * {@link PartitionReader} and combine metrics from partitions to the final result. How Spark
+ * combines metrics depends on the metric type. For streaming query, Spark will collect and combine
+ * metrics for a final result per micro batch.
+ *
+ * The metrics will be gathered during query execution back to the driver and then combined. The
+ * final result will be shown up in the physical operator in Spark UI.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public interface CustomMetric {
+    /**
+     * Returns the name of custom metric.
+     */
+    String name();
+
+    /**
+     * Returns the description of custom metric.
+     */
+    String description();
+
+    /**
+     * Supported metric type. The metric types must be supported by Spark SQL internal metrics.
+     * SUM: Spark sums up metrics from partitions as the final result.
+     */
+    enum MetricType {
+      SUM
```

Comment:
I only use sum metric for now for the Kafka scan purpose. So I leave other possible metric type (size, timing, average) out for now to make it simpler at the beginning.

Reply:
Any reason why we do not allow users to define the combine method instead? The current API means if a user needs to use a different combine behavior, they need to submit a PR to Spark to add it and wait for a new Spark release.

Reply:
By the way, regarding the naming, I would prefer `MergeType` to make it clear if we did go to this direction.

Reply:
Because these metrics are logical representation and are collected to be SQL metrics internally. They are used by DS v2 to report SQL metrics easily. So SQL metrics define how metrics are combined/aggregated. A similar case is public expression APIs for predicate pushdown, as they are converted from catalyst expressions so are matched to catalyst expressions.

Reply:
I'm wondering if we can design a better API. This would be a public API and hard to change in future. We don't have to build it in a way just because of what SQLMetric can do today. When I added `SQLMetric` to Spark SQL, we didn't pay much attention to this since it's an internal API. But making a new public API is a different story and we need to think about how  to use it from the user perspective.

Reply:
Hmm, these metrics are not general metrics that could be defined and used by end users. It sounds like we already have Accumulator for the purpose. Defining a public API of general metrics for end users is overlapping with Accumulator, IMHO.

The metrics here are used by DS v2 implementations to report metrics to Spark SQL. That being said, it is not exposed for end users as general metrics. I think the purpose for the metric API is clear.

Reply:
> Defining a public API of general metrics for end users is overlapping with Accumulator, IMHO.

Totally agree. There would be 3 types of metrics APIs (including the internal SQLMetric) if we added this. That's really confusing to users. Is it possible to make SQLMetric support general Accmulator instead so that we don't need to re-invent a new metric API?

> That being said, it is not exposed for end users as general metrics.

Developers building spark data sources are also critical to the whole Spark ecosystem and these APIs are designed for them. IMHO, we should try our best to build a better API if possible.

Reply:
As you can see, the metric API here is a logical representation of metrics from DS v2. We are not going to re-invent a whole metric API. SQLMetrics are internal to Spark. It is not exposed to end users and data source developers, so I don't think it worries me too much.

I'm not saying that we should not build a good API for DS v2 developers. Seems to me some points in above comments are from end user perspective, I'd like to point out this is for different scenarios.

As this is used for DS v2 purpose, it is for SQL metrics and internally it is converted to SQL metrics. To make SQLMetric support Accmulator and let DS v2 reports Accmulator does not sound bad idea to me. But I'd doubt if it is worth.

One argued point is to define arbitrary combine behavior. Once making SQLMetric support Accmulator, does it mean that we can use arbitrary Accmulator? No, basically SQLMetric allows certain types of metrics, we still need to change SQLMetric to support new metrics.

So the only benefit I thought is to not have another metric API. And I don't think it is serious for this case at the beginning.
 
This API is pretty simple as it is just logical representation and we only need small change internally to convert collected metrics from DS v2 to SQL metrics.

I just read through the code path to be touched in order to make SQLMetric support Accmulator. Seems it involves more changes not only in DS v2 but maybe also in sql/core, etc.

Although I doubt if it is worth, I'm open to the suggested Accmulator approach. Let's gather more thoughts from others?

cc @cloud-fan @dongjoon-hyun @rdblue @Ngone51 @sunchao WDYT?